### DISCUSSION TITLE
No GPU support for mstump

### DISCUSSION THREAD

**Discussion by barroca8** [TDAmeritrade/stumpy#700] [Help: Coding & Implementations]
First of all, congratulations! I am a huge fan of your work in stumpy @seanlaw !

I noticed that there is no GPU support for `mstump` ( https://github.com/TDAmeritrade/stumpy/discussions/262#discussioncomment-196392 ) and I was wondering if there is anything I can do to go around this. 

My data has 3 dimensions in 3 columns ("x", "y", "z") and I was wondering if there was any way to do the equivalent of what would be `stumpy.gpu_mstump` for these 3 dimensions. Example:

`mps, indices = stumpy.gpu_mstump(df, m)`

 The GPU part is really important for me because simply doing this with `mstump` takes me more than 2 hours and eventually crashes the computer.

If I do this process individually does it return the same results? Example:

```
window_size = 1000
mp_x, indices_x = stumpy.gpu_stump(df["x"].astype(float), m=window_size,ignore_trivial=True)
mp_y, indices_y = stumpy.gpu_stump(df["y"].astype(float), m=window_size,ignore_trivial=True)
mp_z, indices_z = stumpy.gpu_stump(df["z"].astype(float), m=window_size,ignore_trivial=True)
```

**Comment by seanlaw** [comment]
@barroca8 Thank you for your question/kinds words and welcome to the STUMPY community!

> I noticed that there is no GPU support for mstump ( https://github.com/TDAmeritrade/stumpy/discussions/262#discussioncomment-196392 ) and I was wondering if there is anything I can do to go around this.

You're correct. There is a long-standing open issue #90 for eventually adding a `gpu_mstump` function but it is non-trivial. PRs are certainly welcome!

> If I do this process individually does it return the same results? 

No, multi-dimensional matrix profiles are **NOT** equivalent to computing individual 1D matrix profiles and stacking them together. I recommend reading the [MSTUMP tutorial](https://stumpy.readthedocs.io/en/latest/Tutorial_Multidimensional_Motif_Discovery.html) in order to understand the difference. Given that you've asked this question, perhaps `mstump` isn't what you need? The interpretation may not be as straightforward as you may think.

> My data has 3 dimensions in 3 columns ("x", "y", "z") and I was wondering if there was any way to do the equivalent of what would be stumpy.gpu_mstump for these 3 dimensions.

Computing multi-dimensional matrix profiles can be very expensive. You may try subsampling your data, say, every 10th or 100th data point instead to reduce the computational time and memory requirements.

**Reply by barroca8** [reply]
Thank you for the quick answer! I think in this case it really is 3 dimensions instead of 3 1D stacked, because this data comes from an accelerometer, measuring the acceleration in X, Y and Z every 2 microseconds. I am using Google Colab for this task because they have a GPU system, but I can try to subsample my data, given that it is more or less 500,000 rows, I can try to reduce it to 50,000 rows or even 5,000 if it's necessary

**Reply by NimaSarajpoor** [reply]
@barroca8 
So, if you have just three dimensions and if  you think all three are meaningful (and I think they are because they are (x,y,z) coordinates of a point in R3 space), then you may want to avoid using `mstump` as what it does is to capture the meaningful dimensions (and as @seanlaw  said, it is computationally expensive). In my opinion, three dimensions is not that much and I think they are all meaningful here.

So, I THINK you should try to modify `stump` or `gpu_stump` so it can calculate the matrix profile by considering all three dimensions together. Modifying `gpu_stump` may not be easy. I think `stump` (no support for GPU) is easier. So, I assume you can reduce `500_000` time stamps to `50_000` time stamps. Then, you can try to modify `stump` or `aamp`:

* if you want to z-normalize each dimension: then, you can modify `stump` to accept three time series `T_Ax`, `T_Ay`, `T_Az`, and then you can do:

Replace 
```
# see line 188 in stumpy/stump.py
pearson = cov * Σ_T_inverse[i + k] * σ_Q_inverse[i]
```
with:
```
pearson_x = cov_x * Σ_T_inverse_x[i + k] * σ_Q_inverse_x[i]
pearson_y = cov_y * Σ_T_inverse_y[i + k] * σ_Q_inverse_y[i]
pearson_z = cov_z * Σ_T_inverse_z[i + k] * σ_Q_inverse_z[i]

pearson = (pearson_x + pearson_y + pearson_z) / 3
```

And I hope you can get the idea here. You first need to repeat `pearson = cov * Σ_T_inverse[i + k] * σ_Q_inverse[i]` for three dimensions and then take their average. So, each time you are updating `pearson` in each dimension, and you do it for all three dimensions. Then, just take their average. 

It can mathematically be proved that: $EDznorm = \sqrt{\frac{EDznorm_{x}^{2} + EDznorm_{y}^{2} + EDznorm\_{z}^{2}}{3}}$ is the same as  $EDznorm$ that corresponds to a pearson value that is average of the three pearson values `pearson_x`, `pearson_y`, and `pearson_z`. However, if you want to combine distances of three dimensions like this: $EDzorm = \frac{EDznorm_{x} + EDznorm_{y} + EDznorm\_{z}}{3}$, then you cannot use the aforementioned trick! So, it depends on how you want to combine the EDznorm distances of the three dimensions. I think using square root of average of squares is reasonable as you are trying to avoid large distance in one dimension. (this is similar to manhattan vs euclidean distance)

* if you do not want to z-normalize subseq before calculating the distance, then just modify `aamp` (non-normalized version of `stump`) and I think it would be less complicated to modify that function.
